### PR TITLE
Do not fall back to shallow copy in defaultDeepCopy

### PR DESCRIPTION
**Pull request by liggitt**
This badly bit us with a custom type that defaulted to shallow copy of a private field. Would vastly have preferred an error indicating a custom DeepCopy() implementation was required.

All the existing types affected by this should have DeepCopy() impls already

- [x] generator PR at https://github.com/kubernetes/gengo/pull/18
- [x] use api.Scheme.DeepCopy instead of instantiating new Cloners that don't have custom/generated deepcopy functions registered
- [x] pass addressable values DeepCopy()

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35728)

### CODE REVIEW

**Comment by liggitt** [comment]
cc @deads2k

**Comment by deads2k** [comment]
reassigned since daniel is out

File: pkg/conversion/cloner.go [hunk 1]
```diff
@@ -228,11 +228,8 @@ func (c *Cloner) defaultDeepCopy(src reflect.Value) (reflect.Value, error) {
 		dst := reflect.New(src.Type())
 		for i := 0; i < src.NumField(); i++ {
 			if !dst.Elem().Field(i).CanSet() {
-				// Can't set private fields. At this point, the
-				// best we can do is a shallow copy. For
-				// example, time.Time is a value type with
-				// private members that can be shallow copied.
-				return src, nil
+				// A custom DeepCopy() implementation is required
+				return src, fmt.Errorf("cannot deep copy field %d of %v", i, src.Type())
```
Comment (smarterclayton) [inline pkg/conversion/cloner.go]: Should have run `rm -rf /` on your system just to teach you a lesson.

**Comment by smarterclayton** [comment]
Test for deep copy somewhere?

**Comment by liggitt** [comment]
> Test for deep copy somewhere?

our whole codebase is a carefully crafted test of deepcopy

**Comment by smarterclayton** [comment]
Haha.  A follow up issue then to add a fuzzing test for deep copy mutation.

**Comment by liggitt** [comment]
cc @sttts

File: pkg/conversion/cloner.go [hunk 2]
```diff
@@ -153,6 +153,36 @@ func (c *Cloner) deepCopy(src reflect.Value) (reflect.Value, error) {
 		err := fv(src.Interface(), outValue.Interface(), c)
 		return outValue, err
 	}
+
+	// Check for custom deep copy funcs on the pointer of this type
+	if src.Kind() == reflect.Struct {
```
Comment (sttts) [inline pkg/conversion/cloner.go]: This case is slow and if the code path is used, it's a (performance) bug in the code. We should at least `glog.Errorf` here. Background: go copies the struct on every function call. This sums up and costs.

Comment (liggitt) [inline pkg/conversion/cloner.go]: a reflect.Value is passed here... we hit this all over the place traversing struct fields (deepCopyDefault calls back here recursively)

**Comment by sttts** [comment]
IMO adding better non-pointer struct support is not the right direction. We should discourage the use of these code paths. Passing (huge) structs costs.

Instead of `glog.Error` (or panic) it would be even better to enforce a pointer type. An ad-hoc idea: `func DeepCopy(*interface{}) *interface{}`.

**Comment by smarterclayton** [comment]
Generated copiers shouldn't be coming into this branch.  Why are they
coming into this branch?

**Comment by liggitt** [comment]
`DeepCopy()` of any object that doesn't have a registered custom or generated copy function (like `DeepCopy(api.Pod{})`) goes into the `defaultDeepCopy(...)` path, which recurses and calls `deepCopy(...)` with all child fields. Even if some of those child types have registered functions, they typically don't get called because `defaultDeepCopy()` frequently passes the non-pointer reflect.Value objects.

**Comment by smarterclayton** [comment]
Why would we have any types without copy functions?  That seems like a bug
in our copy generation (such that if we are going to worry about
performance we should be worrying about fixing that).

**Comment by liggitt** [comment]
We have generated funcs for *api.Pod, not api.Pod (same for all substructs, we generate for pointers, not structs).

That means if you call DeepCopy wrong, you don't get the generated func.

**Comment by smarterclayton** [comment]
Ah.  So that's probably something that should panic during our test runs.

**Comment by liggitt** [comment]
The "Pass addressable values to DeepCopy" commit contains fixes to the places we were trying to copy in ways that would not have picked up the generated copy functions

**Comment by liggitt** [comment]
@wojtek-t @smarterclayton PTAL, especially at the "Pass addressable values to DeepCopy" commit

Doing `conversion.NewCloner().DeepCopy(...)` meant that no generated/custom deepcopy functions were registered in the cloner used to deep copy

Passing non-addressable values to DeepCopy meant a performance hit (both because of the struct copy made to pass to the DeepCopy function, and because no generated deepcopy functions were found for the passed type, since we generate deepcopy functions for pointers to API objects)

In addition to the performance penalty, both cases have the potential to do shallow copies if the reflection-based deep copy hit an unexported field.

**Comment by sttts** [comment]
Here is a WIP PoC to get rid of any reflection and type casts: https://github.com/kubernetes/kubernetes/pull/36412. This should make the deepcopy behavior much simpler.

**Comment by sttts** [comment]
In `federation/pkg/federation-controller/secret/secret_controller.go` is another `NewCloner()` call.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 3]
```diff
@@ -537,7 +536,7 @@ func (ic *IngressController) getMasterCluster() (master *federation_api.Cluster,
 */
 func (ic *IngressController) updateClusterIngressUIDToMasters(cluster *federation_api.Cluster, fallbackUID string) {
 	masterCluster, masterUID, err := ic.getMasterCluster()
-	clusterObj, clusterErr := conversion.NewCloner().DeepCopy(cluster) // Make a clone so that we don't clobber our input param
+	clusterObj, clusterErr := api.Scheme.DeepCopy(cluster) // Make a clone so that we don't clobber our input param
```
Comment (sttts) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Are the ingress deepcopy functions really registered in the global `api.Scheme` object?

File: pkg/runtime/scheme.go [hunk 4]
```diff
@@ -90,6 +90,15 @@ func NewScheme() *Scheme {
 	}
 	s.converter = conversion.NewConverter(s.nameFunc)
 
+	// Add the deep copy funcs that can't auto-register themselves
+	// This should only be the runtime and unversioned packages
+	if err := s.AddGeneratedDeepCopyFuncs(GetGeneratedDeepCopyFuncs()...); err != nil {
+		panic(err)
+	}
+	if err := s.AddGeneratedDeepCopyFuncs(unversioned.GetGeneratedDeepCopyFuncs()...); err != nil {
```
Comment (liggitt) [inline pkg/runtime/scheme.go]: as you can see, no import was added... `runtime` already depends on `unversioned` for GroupVersion and GroupVersionKind types. I actually spent a long time trying to untangle that as part of this, and the ripples were... large. I'd be happy to add a TODO to reverse the direction and expose an unversioned.AddToScheme

**Comment by liggitt** [comment]
@k8s-bot gce node e2e test this

**Comment by liggitt** [comment]
would really like to get this merged... uses of `conversion.NewCloner().DeepCopy` are continuously being added to the code base and copying objects without using generated deep copy funcs. At best, those just perform badly. At worst, they don't actually make copies of objects.

**Comment by liggitt** [comment]
closing in favor of https://github.com/kubernetes/kubernetes/issues/36412

**Pull request closed without merging**